cocotb/handle.py

```diff
@@ -67,18 +67,48 @@ def __init__(self, handle, path):
             path (str): Path to this handle, ``None`` if root.
         """
         self._handle = handle
-        self._len = None
-        self._sub_handles = {}  # Dictionary of children
-        self._invalid_sub_handles = set()  # Set of invalid queries
-
-        self._name = self._handle.get_name_string()
-        self._type = self._handle.get_type_string()
-        self._fullname = self._name + "(%s)" % self._type
-        self._path = self._name if path is None else path
+        self._len = None  # type: int
+        """The "length" (the number of elements) of the underlying object. For vectors this is the number of bits.
+
+        :meta public:
+        """
+        self._sub_handles = {}  # type: dict
+        """Dictionary of this handle's children."""
+        self._invalid_sub_handles = set()  # type: set
+        """Python :class:`set` of invalid queries, for caching purposes."""
+        self._name = self._handle.get_name_string()  # type: str
+        """The name of an object.
+
+        :meta public:
+        """
+        self._type = self._handle.get_type_string()  # type: str
+        """The type of an object as a string.
+
+        :meta public:
+        """
+        self._fullname = self._name + "(%s)" % self._type  # type: str
+        """The name of an object with its type appended in parentheses.
+
+        :meta public:
+        """
+        self._path = self._name if path is None else path  # type: str
+        """The path to this handle, or its name if this is the root handle.
+
+        :meta public:
+        """
         self._log = SimLog("cocotb.%s" % self._name)
+        """The logging object."""
         self._log.debug("Created")
-        self._def_name = self._handle.get_definition_name()
-        self._def_file = self._handle.get_definition_file()
+        self._def_name = self._handle.get_definition_name()  # type: str
+        """The name of a GPI object's definition.
```

Comment:
It would be great to contrast this with `_name` if you can.

Reply:
I added ``_def_name`` and ``_def_file`` to ``test_discovery`` to see what it actually contains, and I found that only for ``HierarchyArrayObject`` the string is not empty: 
```
[...]
     0.00ns INFO     cocotb.sample_module                  test_discovery.py:44   in _discover                       Found name=stream_in_string def_name= def_file= type=(<class 'cocotb.handle.HierarchyObject'>)
     0.00ns INFO     cocotb.sample_module                  test_discovery.py:44   in _discover                       Found name=genblk1 def_name=sample_module def_file=/home/colin/tools/cocotb/tests/designs/sample_module/sample_module.sv type=(<class 'cocotb.handle.HierarchyArrayObject'>)
     0.00ns INFO     cocotb.sample_module                  test_discovery.py:44   in _discover                       Found name=genblk1[0] def_name= def_file= type=(<class 'cocotb.handle.HierarchyObject'>)
[...]
```
I'll open this as another issue in case it's a regression.
This may not be important enough to delay the release, @imphil? I would not make these public then?

Reply:
Ooops, I now realize I don't have the latest code installed, let me try with that.

Reply:
I don't really mind much about this property, it was clearly intended to be public, even if it doesn't work very well.

Reply:
Ok, I'll leave this public, but I'm not sure how to better describe what ``_def_name`` is given the small sample. Minor issue.